kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/RecordTransform.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.records;
+
+import java.nio.ByteBuffer;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.record.Record;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * Represents a transformation from a {@link Record} to (the properties of) a new {@link Record}.
+ * RecordTransforms retain ownership of the {@link ByteBuffer}s returned by their {@code transform*()} methods.
+ * For a given record, once a caller has used those buffers the caller must call
+ * {@link #resetAfterTransform(Record)}.
+ */
+@NotThreadSafe
+public interface RecordTransform {
```

Comment:
This contract is fine for encryption, but it's not quite sufficient for decryption. Specifically in the case where we encrypted both the record value and the record headers and stuffed the resulting wrapper blob in the record value then on decryption we really want to decrypt the blob once and then return the plaintext value and headers from the methods here. The contract currently offered by this interface doesn't define the order in which of the transform*() methods get called, so there is no "one place" to do that decryption.

Reply:
how's init() used? if you require that to called first, isn't that the "one place".

Reply:
Yes. I added `init()` after making this comment, but forgot to dismiss the comment.